components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,699 @@
+            "index" => index,
+            "source_region" => ?source_region,
+        );
+
+        let ctx = TabletContext::new(source_region, None);
+        let source_tablet = reg
+            .tablet_factory()
+            .open_tablet(ctx, &source_path)
+            .unwrap_or_else(|e| {
+                slog_panic!(self.logger, "failed to open source checkpoint"; "err" => ?e);
+            });
+        let open_time = now.saturating_elapsed();
+
+        let mut region = self.region().clone();
+        // Use a max value so that pd can ensure overlapped region has a priority.
+        let version = cmp::max(
+            source_region.get_region_epoch().get_version(),
+            region.get_region_epoch().get_version(),
+        ) + 1;
+        region.mut_region_epoch().set_version(version);
+        if keys::enc_end_key(&region) == keys::enc_start_key(source_region) {
+            region.set_end_key(source_region.get_end_key().to_vec());
+        } else {
+            region.set_start_key(source_region.get_start_key().to_vec());
+        }
+
+        let path = reg.tablet_path(self.region_id(), index);
+        if path.exists() {
+            // Redo merge. Source tablet is flushed before creating this path. The fact that
+            // we need to replay `CommitMerge` means `self.tablet().flush_cfs()` has not
+            // completed. So it's safe to delete it without losing any writes.
+            std::fs::remove_dir_all(&path).unwrap();
+        }
+        // Avoid seqno jump back between self.tablet and the newly created tablet.
+        self.tablet().flush_cfs(&[], true).unwrap();
+        let flush_time = now.saturating_elapsed();
+        let mut ctx = TabletContext::new(&region, Some(index));
+        ctx.flush_state = Some(self.flush_state().clone());
+        let tablet = reg.tablet_factory().open_tablet(ctx, &path).unwrap();
+        tablet
+            .merge(&[&source_tablet, self.tablet()])
+            .unwrap_or_else(|e| {
+                slog_panic!(
+                    self.logger,
+                    "fails to merge tablet";
+                    "path" => %path.display(),
+                    "error" => ?e
+                )
+            });
+        let merge_time = now.saturating_elapsed();
+
+        info!(
+            self.logger,
+            "applied CommitMerge";
+            "source_region" => ?source_region,
+            "wait" => %ReadableDuration(wait_time),
+            "open" => %ReadableDuration(open_time.saturating_sub(wait_time)),
+            "merge" => %ReadableDuration(flush_time.saturating_sub(open_time)),
+            "flush" => %ReadableDuration(merge_time.saturating_sub(flush_time)),
+        );
+
+        self.set_tablet(tablet.clone());
+
+        self.region_state_mut().set_region(region.clone());
+        self.region_state_mut().set_state(PeerState::Normal);
+        self.region_state_mut().clear_merge_state();
+        self.region_state_mut().set_tablet_index(index);
+
```

Comment:
Missing updating `removed_records` and `merged_records`.

Reply:
New leader takes the responsibility to clean up removed peer of source region.

Reply:
I mean, isn't that written to `merged_records`? Why write again to `removed_records`?

Reply:
There are also `removed_records` in source region. `removed_records` and `merged_records` need to be merged into target region respectively. And all peers of source region should go to `merged_records`.